library/core/src/ptr/mod.rs

```diff
@@ -777,8 +777,8 @@ where
 
 /// Convert a reference to a raw pointer.
 ///
-/// This is equivalent to `r as *const T`, but is a bit safer since it will never silently change
-/// type or mutability, in particular if the code is refactored.
+/// For `r: &T`, `from_ref(r)` is equivalent to `r as *const T`, but is a bit safer since it will
+/// never silently change type or mutability, in particular if the code is refactored.
```

Comment:
In addition, I think it is worth also clarifying that for `r: &mut T`, `from_ref(r)` is equivalent to `from_ref(r as &T)` which is *not* guaranteed to be equivalent to `r as *const` and is *not*  necessarily safer than `r as const T`, in particular if the resulting pointer will ever be cast to a `*mut _` pointer. Maybe there is similar nuance for `r: Deref<T>`.

However, https://doc.rust-lang.org/reference/type-coercions.html#coercion-types does seem to say that when `r: &mut T`, `r as *const T` is equivalent to `(r as &T) as *const T` since there is no coercion rule for `&mut T to *const T`. So first we should decide whether we can really make the potentially-backward-incompatible change to add a `&mut T` to `*const T` coercion rule, because if we decide we can't, then this new wording isn't necessary.

Reply:
>  for `r: &mut T`, `r as *const T` is equivalent to `(r as &T) as *const T` since there is no coercion rule for `&mut T to *const T`. 

Actually, `r as *const T` could also be `(r as *mut T) as *const T`, so I guess some clarification is needed.

Reply:
I don't think this is the right place to discuss the semantics of coercions. They behave the same everywhere, surely we don't want to explain them in each function that takes a reference. (e.g. `slice.as_ptr()` has exactly the same questions).

Reply:
It's analogous to the situation for `addr_of!`, which has a similar caveat:
> Note that Deref/Index coercions (and their mutable counterparts) are applied inside `addr_of_mut!` like everywhere else, in which case a reference is created to call `Deref::deref` or `Index::index`, respectively. The statements above only apply when no such coercions are applied.

In both cases, the construct is designed to make things safer, but not understanding the interaction with coercions is especially dangerous.

Reply:
I suggest adding the following text, which really gets to the point:

For `r: &mut T`, instead use `from_mut(r).cast_const()`, instead of `from_ref(r)`, to get a `*const T` that can later safely be cast to `*mut T`.